File: JAVA/236 - Lowest Common Ancestor of a Binary Tree.java
/*
 236. Lowest Common Ancestor of a Binary Tree (M)

 Approach 1 : Postorder traversal
 If find p or q, return p or q to parent.
 If find both p and q from parent's children, return that parent
 Time: O(n)
 Space: O(n)

*/

/* Approach 1 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        
        if (root == p || root == q) return root;
        if (left != null && right != null) return root;
        
        return left == null ? right : left;
    }
}

File: C++/236-Lowest_Common_Ancestor_of_a_Binary_Tree.cpp

/*
 236. Lowest Common Ancestor of a Binary Tree (M)

 Notes:
 Approach 1 : Postorder traversal
 Use Postorder traversal. If found, return to parent. 
 Time: O(n), traverse all nodes
 Space: O(n)

*/

// Approach 1 : Postorder traversal
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root)
            return nullptr;

        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);

        if (root == p || root == q)
            return root;

        if (left && right)
            return root;

        return left == nullptr ? right : left;
    }
};